constraints/match.js

```diff
@@ -111,5 +137,93 @@ function hasMatch(sNode, pTree)
 		}
 		return false;
 	}
-	
+
+}
+
+// Match Max constraints:
+
+/* Same as hasMatch function above, except this only returns true if the
+ * matching prosodic node is maximal:
+ */
+function hasMaxMatch(sNode, pTree){
+	 var sLeaves = getLeaves(sNode);
+	 markMinMax(pTree); //mark min and max on prosodic tree
+	 if(catsMatch(sNode.cat, pTree.cat) && sameIds(getLeaves(pTree), sLeaves) && pTree.isMax)
+	 // the current prosodic node is the match, both for category and for terminals, and is maximal
+		{
+ 			return true;
+ 		}
+
+ 		// If the current prosodic node is NOT the match:
+
+ 		else if(!pTree.children || pTree.children.length===0){
+ 		// current node is terminal
+ 			return false;
+		}
+
+ 		else
+ 		//the current prosodic node is non-terminal (has children)
+ 		{
+ 			for(var i = 0; i < pTree.children.length; i++)
+ 			//check each child to see if the match exists in the subtree rooted in that child
+ 			{
+ 				var child = pTree.children[i];
+ 				if(hasMaxMatch(sNode, child)){
+ 					return true;
+				}
+ 			}
+ 			return false;
+ 		}
+
+ }
+
+ /* Assign a violation for every node of syntactic category s that is not
+  * dominated by another node of category s in the syntactic tree, and is not
+  * mapped to a corresponding prosodic node of category p, where p=catMap(s),
+  * such that p is not dominated by another node of category p.
+  */
+
+function matchMaxSP(sTree, pTree, sCat){
+	 var vcount = 0;
+	 markMinMax(sTree); //mark maximal nodes in tree
+	 if (sTree.children && sTree.children.length){
+		 for (var i = 0; i < sTree.children.length; i ++){
+			 vcount += matchMaxSP(sTree.children[i], pTree, sCat); //recursive function call
+		 }
+	 }
+	 if (sTree.cat === sCat && sTree.isMax && !hasMaxMatch(sTree, pTree)){
+		 //add violation if this node has no maximal match, is maximal and of the right cat
+		 vcount ++;
+	 }
+	 return vcount;
+ }
+
+/* Same as matchMaxSP, except matching prosodic node need not be maximal,
```

Comment:
Just for clarification, please add the following to the comment here:
Match-SP(scat-max, pcat)
ex. Match a maximal xp with any phi

Reply:
added in latest commit.